server/src/main/java/org/opensearch/cluster/health/ClusterIndexHealth.java

```diff
@@ -189,6 +189,102 @@ public ClusterIndexHealth(final IndexMetadata indexMetadata, final IndexRoutingT
         this.relocatingShards = computeRelocatingShards;
         this.initializingShards = computeInitializingShards;
         this.unassignedShards = computeUnassignedShards;
+        this.delayedUnassignedShards = computeDelayedUnassignedShards;
+    }
+
+    public ClusterIndexHealth(
+        final IndexMetadata indexMetadata,
+        final IndexRoutingTable indexRoutingTable,
+        final ClusterHealthRequest.Level healthLevel
+    ) {
+        this.index = indexMetadata.getIndex().getName();
+        this.numberOfShards = indexMetadata.getNumberOfShards();
+        this.numberOfReplicas = indexMetadata.getNumberOfReplicas();
+
+        shards = new HashMap<>();
+
+        // update the index status
+        ClusterHealthStatus computeStatus = ClusterHealthStatus.GREEN;
+        int computeActivePrimaryShards = 0;
+        int computeActiveShards = 0;
+        int computeRelocatingShards = 0;
+        int computeInitializingShards = 0;
+        int computeUnassignedShards = 0;
+        int computeDelayedUnassignedShards = 0;
+
+        boolean isShardLevelHealthInfo = healthLevel == ClusterHealthRequest.Level.SHARDS;
+        if (!isShardLevelHealthInfo) {
+            for (IndexShardRoutingTable indexShardRoutingTable : indexRoutingTable) {
+                int activeShardsPerShardId = 0;
+
+                List<ShardRouting> shardRoutings = indexShardRoutingTable.shards();
+                int shardRoutingCountPerShardId = shardRoutings.size();
+                for (int index = 0; index < shardRoutingCountPerShardId; index++) {
+                    ShardRouting shardRouting = shardRoutings.get(index);
+                    if (shardRouting.active()) {
+                        computeActiveShards++;
+                        activeShardsPerShardId++;
+                    } else if (shardRouting.relocating()) {
+                        computeRelocatingShards++;
+                    } else if (shardRouting.initializing()) {
+                        computeInitializingShards++;
+                    } else if (shardRouting.unassigned()) {
+                        computeUnassignedShards++;
+                        if (shardRouting.unassignedInfo().isDelayed()) {
+                            computeDelayedUnassignedShards++;
+                        }
+                    }
+                }
```

Comment:
all this code will be duplicated across ClusterIndexHealth & ClusterShardHealth?

Reply:
We should think about keeping the information calculated as part of IndexRoutingTable using IndexShardRoutingTable itself. It can serve as cache as well. We should skip this intermediate step to just optimize ClusterStateHealth object instantiation

Reply:
This duplication is arising from how the data models are defined and couldn't have been avoided without introducing some other abstract classes like cache, this requires holistic thinking to weigh its benefits as it will have its own memory foot-print and comes with challenge's of keeping the cache upto date. 

Given the performance benefits we have observed with the in-line aggregations, do you think caching change can be done incrementally?

Reply:
Git hub issue created for exploring cache based solution: https://github.com/opensearch-project/OpenSearch/issues/15685